File: .NET/Microsoft.Recognizers.Text.DateTime/French/Extractors/FrenchTimeExtractorConfiguration.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;

using Microsoft.Recognizers.Definitions.French;
using Microsoft.Recognizers.Definitions.Utilities;

namespace Microsoft.Recognizers.Text.DateTime.French
{
    public class FrenchTimeExtractorConfiguration : BaseDateTimeOptionsConfiguration, ITimeExtractorConfiguration
    {
        // part 1: smallest component
        // --------------------------------
        public static readonly Regex DescRegex =
            RegexCache.Get(DateTimeDefinitions.DescRegex, RegexFlags);

        public static readonly Regex HourNumRegex =
            RegexCache.Get(DateTimeDefinitions.HourNumRegex, RegexFlags);

        public static readonly Regex MinuteNumRegex =
            RegexCache.Get(DateTimeDefinitions.MinuteNumRegex, RegexFlags);

        // part 2: middle level component
        // --------------------------------------
        // handle "... heures (o'clock, en punto)"
        public static readonly Regex OclockRegex =
            RegexCache.Get(DateTimeDefinitions.OclockRegex, RegexFlags);

        // handle "... après midi (afternoon, tarde)"
        public static readonly Regex PmRegex =
            RegexCache.Get(DateTimeDefinitions.PmRegex, RegexFlags);

        // handle "... dans la matinee (in the morning)"
        public static readonly Regex AmRegex =
            RegexCache.Get(DateTimeDefinitions.AmRegex, RegexFlags);

        // handle "half past ..." "a quarter to ..."
        // rename 'min' group to 'deltamin'
        public static readonly Regex LessThanOneHour =
            RegexCache.Get(DateTimeDefinitions.LessThanOneHour, RegexFlags);

        public static readonly Regex WrittenTimeRegex =
            RegexCache.Get(DateTimeDefinitions.WrittenTimeRegex, RegexFlags);

        // TODO - will have change below
        // handle "six heures et demie" (six thirty), "six heures et vingt-et-un" (six twenty one)
        public static readonly Regex TimePrefix =
            RegexCache.Get(DateTimeDefinitions.TimePrefix, RegexFlags);

        public static readonly Regex TimeSuffix =
            RegexCache.Get(DateTimeDefinitions.TimeSuffix, RegexFlags);

        public static readonly Regex BasicTime =
            RegexCache.Get(DateTimeDefinitions.BasicTime, RegexFlags);

        // handle special time such as 'at midnight', 'midnight', 'midday'
        // midnight - le minuit, la zero heure
        // midday - midi
        public static readonly Regex MidnightRegex =
            RegexCache.Get(DateTimeDefinitions.MidnightRegex, RegexFlags);

        public static readonly Regex MidmorningRegex =
            RegexCache.Get(DateTimeDefinitions.MidmorningRegex, RegexFlags);

        public static readonly Regex MidafternoonRegex =
            RegexCache.Get(DateTimeDefinitions.MidafternoonRegex, RegexFlags);

        public static readonly Regex MiddayRegex =
            RegexCache.Get(DateTimeDefinitions.MiddayRegex, RegexFlags);

        public static readonly Regex MidTimeRegex =
            RegexCache.Get(DateTimeDefinitions.MidTimeRegex, RegexFlags);

        // part 3: regex for time
        // --------------------------------------
        // handle "at four" "at 3"
        public static readonly Regex AtRegex =
            RegexCache.Get(DateTimeDefinitions.AtRegex, RegexFlags);

        public static readonly Regex IshRegex =
            RegexCache.Get(DateTimeDefinitions.IshRegex, RegexFlags);

        public static readonly Regex TimeUnitRegex =
            RegexCache.Get(DateTimeDefinitions.TimeUnitRegex, RegexFlags);

        public static readonly Regex ConnectNumRegex =
            RegexCache.Get(DateTimeDefinitions.ConnectNumRegex, RegexFlags);

        public static readonly Regex TimeBeforeAfterRegex =
            RegexCache.Get(DateTimeDefinitions.TimeBeforeAfterRegex, RegexFlags);

        public static readonly Regex[] TimeRegexList =
        {
            // (three min past)? seven|7|(seven thirty) pm
            RegexCache.Get(DateTimeDefinitions.TimeRegex1, RegexFlags),

            // (three min past)? 3:00(:00)? (pm)?
            RegexCache.Get(DateTimeDefinitions.TimeRegex2, RegexFlags),

            // (three min past)? 3.00 (pm)?
            RegexCache.Get(DateTimeDefinitions.TimeRegex3, RegexFlags),

            // (three min past) (five thirty|seven|7|7:00(:00)?) (pm)? (in the night)
            RegexCache.Get(DateTimeDefinitions.TimeRegex4, RegexFlags),

            // (three min past) (five thirty|seven|7|7:00(:00)?) (pm)?
            RegexCache.Get(DateTimeDefinitions.TimeRegex5, RegexFlags),

            // (five thirty|seven|7|7:00(:00)?) (pm)? (in the night)
            RegexCache.Get(DateTimeDefinitions.TimeRegex6, RegexFlags),

            // (in the night) at (five thirty|seven|7|7:00(:00)?) (pm)?
            RegexCache.Get(DateTimeDefinitions.TimeRegex7, RegexFlags),

            // (in the night) (five thirty|seven|7|7:00(:00)?) (pm)?
            RegexCache.Get(DateTimeDefinitions.TimeRegex8, RegexFlags),

            RegexCache.Get(DateTimeDefinitions.TimeRegex9, RegexFlags),

            RegexCache.Get(DateTimeDefinitions.TimeRegex10, RegexFlags),

            // 340pm
            ConnectNumRegex,
        };

        private const RegexOptions RegexFlags = RegexOptions.Singleline | RegexOptions.ExplicitCapture;

        public FrenchTimeExtractorConfiguration(IDateTimeOptionsConfiguration config)
            : base(config)
        {
            DurationExtractor = new BaseDurationExtractor(new FrenchDurationExtractorConfiguration(this));
            TimeZoneExtractor = new BaseTimeZoneExtractor(new FrenchTimeZoneExtractorConfiguration(this));
        }

        Regex ITimeExtractorConfiguration.IshRegex => IshRegex;

        IEnumerable<Regex> ITimeExtractorConfiguration.TimeRegexList => TimeRegexList;

        Regex ITimeExtractorConfiguration.AtRegex => AtRegex;

        Regex ITimeExtractorConfiguration.TimeBeforeAfterRegex => TimeBeforeAfterRegex;

        public IDateTimeExtractor DurationExtractor { get; }

        public IDateTimeExtractor TimeZoneExtractor { get; }

        public string TimeTokenPrefix => DateTimeDefinitions.TimeTokenPrefix;

        public Dictionary<Regex, Regex> AmbiguityFiltersDict => DefinitionLoader.LoadAmbiguityFilters(DateTimeDefinitions.AmbiguityTimeFiltersDict);
    }
}

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/french/extractors/FrenchTimeExtractorConfiguration.java
package com.microsoft.recognizers.text.datetime.french.extractors;

import com.microsoft.recognizers.text.datetime.DateTimeOptions;
import com.microsoft.recognizers.text.datetime.config.BaseOptionsConfiguration;
import com.microsoft.recognizers.text.datetime.extractors.BaseDurationExtractor;
import com.microsoft.recognizers.text.datetime.extractors.BaseTimeZoneExtractor;
import com.microsoft.recognizers.text.datetime.extractors.IDateTimeExtractor;
import com.microsoft.recognizers.text.datetime.extractors.config.ITimeExtractorConfiguration;
import com.microsoft.recognizers.text.datetime.resources.FrenchDateTime;
import com.microsoft.recognizers.text.utilities.RegExpUtility;
import java.util.ArrayList;
import java.util.regex.Pattern;

public class FrenchTimeExtractorConfiguration extends BaseOptionsConfiguration implements ITimeExtractorConfiguration {

    public static final Pattern DescRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.DescRegex);
    public static final Pattern HourNumRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.HourNumRegex);
    public static final Pattern MinuteNumRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.MinuteNumRegex);

    public static final Pattern OclockRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.OclockRegex);
    public static final Pattern PmRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.PmRegex);
    public static final Pattern AmRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.AmRegex);

    public static final Pattern LessThanOneHour = RegExpUtility.getSafeRegExp(FrenchDateTime.LessThanOneHour);
    //     public static final Pattern TensTimeRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.TensTimeRegex);

    public static final Pattern WrittenTimeRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.WrittenTimeRegex);
    public static final Pattern TimePrefix = RegExpUtility.getSafeRegExp(FrenchDateTime.TimePrefix);
    public static final Pattern TimeSuffix = RegExpUtility.getSafeRegExp(FrenchDateTime.TimeSuffix);
    public static final Pattern BasicTime = RegExpUtility.getSafeRegExp(FrenchDateTime.BasicTime);
    public static final Pattern IshRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.IshRegex);

    public static final Pattern AtRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.AtRegex);
    public static final Pattern ConnectNumRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.ConnectNumRegex);
    public static final Pattern TimeBeforeAfterRegex = RegExpUtility.getSafeRegExp(FrenchDateTime.TimeBeforeAfterRegex);
    public static final Iterable<Pattern> TimeRegexList = new ArrayList<Pattern>() {
        {
            add(RegExpUtility.getSafeRegExp(FrenchDateTime.TimeRegex1));
            add(RegExpUtility.getSafeRegExp(FrenchDateTime.TimeRegex2));
            add(RegExpUtility.getSafeRegExp(FrenchDateTime.TimeRegex3));
            add(RegExpUtility.getSafeRegExp(FrenchDateTime.TimeRegex4));
            add(RegExpUtility.getSafeRegExp(FrenchDateTime.TimeRegex5));
            add(RegExpUtility.getSafeRegExp(FrenchDateTime.TimeRegex6));
            add(RegExpUtility.getSafeRegExp(FrenchDateTime.TimeRegex7));
            add(RegExpUtility.getSafeRegExp(FrenchDateTime.TimeRegex8));
            add(RegExpUtility.getSafeRegExp(FrenchDateTime.TimeRegex9));
            add(RegExpUtility.getSafeRegExp(FrenchDateTime.TimeRegex10));
            add(ConnectNumRegex);
        }
    };
    private final IDateTimeExtractor durationExtractor;
    private final IDateTimeExtractor timeZoneExtractor;

    public FrenchTimeExtractorConfiguration() {
        this(DateTimeOptions.None);
    }

    public FrenchTimeExtractorConfiguration(final DateTimeOptions options) {
        super(options);
        durationExtractor = new BaseDurationExtractor(new FrenchDurationExtractorConfiguration());
        timeZoneExtractor = new BaseTimeZoneExtractor(new FrenchTimeZoneExtractorConfiguration(options));
    }

    public final Pattern getIshRegex() {
        return IshRegex;
    }

    public final Iterable<Pattern> getTimeRegexList() {
        return TimeRegexList;
    }

    public final Pattern getAtRegex() {
        return AtRegex;
    }

    public final Pattern getTimeBeforeAfterRegex() {
        return TimeBeforeAfterRegex;
    }

    public final IDateTimeExtractor getDurationExtractor() {
        return durationExtractor;
    }

    public final IDateTimeExtractor getTimeZoneExtractor() {
        return timeZoneExtractor;
    }
}
